process: add reportError

benjamingr opened the pull request:
Fixes: https://github.com/nodejs/node/issues/38947 

I'm opening as draft for discussion - mostly I want feedback on "are we sure crash on next tick" is the right behavior :] (I think it is but want opinions).

This is missing the wpts but that probably shouldn't block landing.

cc @mcollina @addaleax @linkgoron

aduh95 commented on lib/internal/bootstrap/node.js:
```diff
@@ -426,5 +436,11 @@ function setupBuffer() {
       writable: true,
       configurable: true,
     },
+    'reportError': {
+      value: reportError,
+      enumerable: false,
+      writable: true,
+      configurable: true
+    },
```
This should probably go in `lib/internal/bootstrap/browser.js` instead.

benjamingr replied:
Sure, no strong feelings about it but wdyt about the semantics/feature itself?

benjamingr replied:
It would make sense to also move atob/btoa (in a separate PR)?

aduh95 commented on doc/api/globals.md:
```diff
@@ -598,13 +598,30 @@ The object that acts as the namespace for all W3C
+added: REPLACEME
+-->
+
+<!-- type=global -->
```
I think we should land this as experimental given that it's not obvious which behavior it should follow.
```suggestion
> Stability: 1 - Experimental.

<!-- type=global -->
```

aduh95 reviewed (commented):
I'd be +1 on this landing in Node.js as a global. Node.js core it self uses similar utility APIs already, here's a few occurrences I could find:

https://github.com/nodejs/node/blob/43371dd5d2b7c3586e5f41bc30c9cd21bb000d5f/lib/internal/event_target.js#L915-L917
https://github.com/nodejs/node/blob/c12db606eea145d62edff840c277023971bc6e81/lib/internal/debugger/inspect.js#L177
https://github.com/nodejs/node/blob/09c9e5dea42fd11606f0d8988462e291cbfafd32/lib/internal/process/warning.js#L145-L151
https://github.com/nodejs/node/blob/29f714b152434e1dbcf7f78d8877543f2ec51340/lib/internal/webstreams/adapters.js#L290-L295

bmeck commented on lib/internal/bootstrap/node.js:
```diff
@@ -387,6 +387,16 @@ function setupGlobalProxy() {
+    throw new codes.ERR_MISSING_ARGS('error');
+  }
+  queueMicrotask(() => {
+    throw error;
```
should this generate a stack trace if no error was provided? mainly thinking of what happens / how a user could back track the problem from `reportError()`.

benjamingr replied:
Maybe emit a warning (synchronously) with a stack trace if a non-Error was passed?

bmeck replied:
sure

bmeck commented on doc/api/globals.md:
```diff
@@ -598,13 +598,30 @@ The object that acts as the namespace for all W3C
+* `error` {Error|any} an error to report.
+
+The `reportError(error)` method triggers an uncaught exception with the error
+`error`. This will crash the Node.js process by default.
```
might be good to link to uncaught exception handling somewhere here so people can take action on the "will crash" bit of the doc.

benjamingr commented on test/common/index.js:
```diff
@@ -268,6 +268,7 @@ let knownGlobals = [
   setInterval,
   setTimeout,
   queueMicrotask,
+  reportError,
```
@aduh95 I don't understand why - this PR will land on the next semver-major - other versions won't have this in common/index.js so they won't expect the global and will (rightfully) error if it's present right?

aduh95 replied:
IIRC that’s there so `make test-doc` can pass without needing to compile the latest `master`.

Mesteery commented on doc/api/globals.md:
```diff
@@ -598,13 +598,32 @@ The object that acts as the namespace for all W3C
 [`MessageChannel`]: worker_threads.md#class-messagechannel
 [`MessageEvent`]: https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent/MessageEvent
 [`MessagePort`]: worker_threads.md#class-messageport
+[`reportError`][]: https://developer.mozilla.org/en-US/docs/Web/API/reportError
```
```suggestion
[`reportError`]: https://developer.mozilla.org/en-US/docs/Web/API/reportError
```

Mesteery commented on test/parallel/test-process-reportError.mjs:
```diff
@@ -0,0 +1,11 @@
+import { mustCall, mustNotCall } from '../common/index.mjs';
+import { strictEqual } from 'assert';
+{
+  // Verify uncaughtException is called
+  process.on('unhandledRejection', mustNotCall());
+  const e = new Error('boom');
+  process.on('uncaughtException', mustCall((error) => {
+    strictEqual(error, e);
+  }));
+  reportError(e);
+}
```
```suggestion

// Verify uncaughtException is called
process.on('unhandledRejection', mustNotCall());
const e = new Error('boom');
process.on('uncaughtException', mustCall((error) => {
  strictEqual(error, e);
}));
reportError(e);
```
nit

benjamingr replied:
I'll add more tests to justify the scoping :)

benjamingr commented:
Addressed comments @bmeck @Mesteery @aduh95 (I will refactor the internal code to use this in a later PR).

WDYT? Is the semantics reasonable to the point I should move this out of draft?

jasnell commented on lib/internal/bootstrap/browser.js:
```diff
@@ -71,6 +71,30 @@ defineOperation(globalThis, 'setTimeout', timers.setTimeout);
+      // The user called reportError with a non-error so we want to give them
+      // a way to get a stack trace. We do this by emitting a warning the user
+      // so the user can run node with `--trace-warnings` to get a stack tace.
+      process.emitWarning('reportError was called with a non-error.');
```
hmm.... I'm not convinced the warning is necessary or all that useful. Not blocking however.

benjamingr replied:
I mostly thing we shouldn’t encourage users to error with things without a stack trace

ljharb replied:
Does the browser warn on this?

Not every error condition needs a stack trace.

benjamingr replied:
> Not every error condition needs a stack trace.

I am pretty sure when debugging a warning I would __really__ want a  stack trace though note this is opt in with `--trace-warnings` right?

ljharb replied:
The individual user’s desire there may be for a stack trace, but that’s not universal, or the language wouldn't let you throw something without one.

benjamingr replied:
@ljharb as far as I'm aware the language is not _yet_ aware of stack traces anyway? So the language is completely oblivious to the debugging experience (or the user) isn't it?

The only reason stuff like `Error` even exists is that language itself can throw stuff that subclasses it (like RangeError or TypeError) and as a possible extensibility point. The language is specific about this:

> Instances of Error objects are thrown as exceptions when runtime errors occur. The Error objects may also serve
as base objects for user-defined exception classes.

(exists since es3)

In any case aren't "what errors users get" and "what the language lets you do" here completely orthogonal since it's something the runtime and not the language typically does?

ljharb replied:
No - users (should) expect to be able to throw and catch any kind of value, because every kind of value can be an error. Thus, a warning here on an otherwise-valid error will be surprising and confusing to developers who expect that explicitly passing a non-Error object to reportError will be calmly accepted.

aduh95 commented:
This needs a rebase (Node.js 18.x semver cutoff is next week).

devinrhode2 commented:
> I am honestly _still_ kind of not sure we should actually do this - waiting for more people to ask for this or to weigh in this is a good idea

I would fricken love this

camsteffen commented:
I feel like there's a catch-22 here. `reportError` to me sounds a lot like it will _not_ crash. But on the other hand, crashing on unexpected errors is generally the thing to do in node. So maybe this function (name) only makes sense in browsers.
